### PR TITLE
Construction of cloudcore upgrading capability

### LINKED ISSUE #5131: Construction of cloudcore upgrading capability
**What would you like to be added/modified**:

Add upgrade command for the cloud in the keadm.

**Why is this needed**:

We can easy upgrade cloud components through the keadm.

**Before we do that, we need to sort out some questions：**

1. Whether the cloud and edge share the upgrade command `keadm upgrade`, like reset?
  I don't really want to use the existence of a edgecore.yaml file to determine whether it's the edge or the cloud. I prefer to use three-level commands `keadm upgrade cloud` and `keadm upgrade edge`.

1. Do we want to optimize the `init` args?
  There is now some confusion between the args of the `init` command and the args of the `generate`(manifest) command, they reuse some args but not all of them, but they share the same options struct `InitOptions`, I think the options struct need to be defined separetely. E.g.:
    ```golang
    cmd.Flags().StringVar(&initOpts.Manifests, types.Manifests, initOpts.Manifests,
        "Allow appending file directories of k8s resources to keadm, separated by commas")
    cmd.Flags().StringVarP(&initOpts.Manifests, types.Files, "f", initOpts.Manifests,
      "Allow appending file directories of k8s resources to keadm, separated by commas")
    ```

1. I'm not sure what the args `external-helm-root` and `profile` do.

**Cloud upgrade proposal**

- Define the cloud upgrade command args:
    ```golang
    type upgradeOptions struct {
        KubeConfig string
        AdvertiseAddress string // Is it need? We need reset the certificate after the modification
        KubeEdgeVersion string
        Namespace string
        Sets []string
        Profile string  // Is it need?
        ExternalHelmRoot string  // Is it need?
        Force bool
        DryRun bool
    }
    ```
- Merge with newly set values using the installed helm release values.
- Reuse the struct `KubeCloudHelmInstTool` function `runHelmInstall` to execute the helm upgrade.

### PR DESCRIPTION
**Pull request by WillardHu**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

We can easy upgrade cloud components through the keadm.

**Which issue(s) this PR fixes**:

Fixes #5131

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kubeedge/kubeedge/common/constants"
+	types "github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/common"
+	"github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/util"
+	kecharts "github.com/kubeedge/kubeedge/manifests"
+)
+
+const (
+	CloudCoreHelmComponent = "cloudcore"
```
Comment (ZhengXinwei-F) [inline keadm/cmd/keadm/app/cmd/helm/cloudcore.go:34]: It seems that this CloudCoreHelmComponent is an internal constant, please change it to cloudCoreHelmComponent. In addition, global constants should be placed in the **api** folder.

Check other constants as well.

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+		Common: common,
+	}
+}
+
+// Install uses helm client to install cloudcore release
+func (c *CloudcoreHelmTool) Install(opts *types.InitOptions) error {
+	// The flag --force would not care about whether the cloud components exist or not also.
+	// If gives a external helm root, no need to check and verify, because it is always not a cloudcore.
+	if !opts.Force && opts.ExternalHelmRoot == "" {
+		if err := c.verifyCloudcoreOrocessRunning(); err != nil {
```
Comment (ZhengXinwei-F) [inline keadm/cmd/keadm/app/cmd/helm/cloudcore.go:77]: It should be verifyCloudcoreProcessRunning. :)

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+	}
+	if cloudcoreRunning {
+		return fmt.Errorf("the cloudcore is already running on this node, please run reset to clean up first")
+	}
+	return nil
+}
+
+// setImageTagOfComponents when the component image tag is not set, it is set to a default value
+func setImageTagOfComponents(version string, opts types.CloudInitUpdateBase) {
+	if !opts.HasSets(setsKeyCloudcoreImageTag) {
```
Comment (ZhengXinwei-F) [inline keadm/cmd/keadm/app/cmd/helm/cloudcore.go:227]: This block can use array `[]string{setsKeyCloudcoreImageTag,setsKeyControllerManagerImageTag,...}` for for loop operation

**Review by ZhengXinwei-F — COMMENTED** [review]
Can I continue the code review tomorrow?

File: keadm/cmd/keadm/app/cmd/common/types.go [hunk 4]
```diff
@@ -17,37 +17,67 @@ limitations under the License.
...
+
+// GetValidSets returns a valid sets, if the item is an invalid key-value,
+// it is removed from the sets and print the error message.
+func (b CloudInitUpdateBase) GetValidSets() []string {
+	if b.Sets == nil {
+		return nil
+	}
+	res := make([]string, 0, len(b.Sets))
+	for _, s := range b.Sets {
+		p := strings.Split(s, "=")
```
Comment (ZhengXinwei-F) [inline keadm/cmd/keadm/app/cmd/common/types.go:45]: Should valid items be in the "key=value" format? If that's the case, `!=` should be used here.

File: keadm/cmd/keadm/app/cmd/common/types.go [hunk 5]
```diff
@@ -17,37 +17,67 @@ limitations under the License.
...
+		}
+		res = append(res, s)
+	}
+	return res
+}
+
+// HasSets returns the key is in the sets
+func (b CloudInitUpdateBase) HasSets(key string) bool {
+	for _, kv := range b.Sets {
+		if strings.HasPrefix(kv, key) {
```
Comment (ZhengXinwei-F) [inline keadm/cmd/keadm/app/cmd/common/types.go:58]: `hasPrefix` may cause incorrect matching. For instance, suppose the existing key is "edgeNode" and the key you want to obtain is edge.

File: keadm/cmd/keadm/app/cmd/edge/upgrade.go [hunk 6]
```diff
@@ -53,12 +53,14 @@ func NewEdgeUpgrade() *cobra.Command {
...
-		Short: "Upgrade edge component. Upgrade the edge node to the desired version.",
-		Long:  "Upgrade edge component. Upgrade the edge node to the desired version.",
+		Use:   "edge",
+		Short: "Upgrade edge components.",
+		Long:  "Upgrade edge components. Upgrade the edge node to the desired version.",
 		RunE: func(cmd *cobra.Command, args []string) error {
 			// upgrade edgecore
-			return upgradeOptions.upgrade()
+			klog.Info("upgrading edge components...")
+			return nil
```
Comment (ZhengXinwei-F) [inline keadm/cmd/keadm/app/cmd/edge/upgrade.go:62]: Is this a debug modification?

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+If you want to revert configuration items, please manually modify the configmap 'cloudcore' and restart the cloudcore: \n%s`
+)
+
+// CloudcoreHelmTool a cloudcore helm chart operation wrapping tool
+type CloudcoreHelmTool struct {
+	util.Common
+}
+
+// NewCloudcoreHelmTool creates a new instance of CloudcoreHelmTool
+func NewCloudcoreHelmTool(kubeConfig, kubeedgeVersion string) *CloudcoreHelmTool {
```
Comment (ZhengXinwei-F) [inline keadm/cmd/keadm/app/cmd/helm/cloudcore.go:60]: It should be NewCloudCoreHelmTool.

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+STATUS: %s
+REVISION: %d
+`
+
+	MessageFormatUpgradationPrintConfig = `This is cloudcore configuration of the previous version.
+If you want to revert configuration items, please manually modify the configmap 'cloudcore' and restart the cloudcore: \n%s`
+)
+
+// CloudcoreHelmTool a cloudcore helm chart operation wrapping tool
+type CloudcoreHelmTool struct {
```
Comment (ZhengXinwei-F) [inline keadm/cmd/keadm/app/cmd/helm/cloudcore.go:55]: CloudCoreHelmTool. :-)

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+	common.OSTypeInstaller.SetKubeEdgeVersion(common.ToolVersion)
+	return &CloudcoreHelmTool{
+		Common: common,
+	}
+}
+
+// Install uses helm client to install cloudcore release
+func (c *CloudcoreHelmTool) Install(opts *types.InitOptions) error {
+	// The flag --force would not care about whether the cloud components exist or not also.
+	// If gives a external helm root, no need to check and verify, because it is always not a cloudcore.
```
Comment (ZhengXinwei-F) [inline keadm/cmd/keadm/app/cmd/helm/cloudcore.go:75]: I'm not sure what the purpose of utilizing ExternalHelmRoot was before, but it was not in the development process this time. You can keep it in the installation procedure but exclude it from the upgrade process.

**Review by ZhengXinwei-F — COMMENTED** [review]
I noticed that some of the code was not debugged; perhaps I should keep reviewing it once you have debugged it.

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go (hunk 1 above)
Comment (WillardHu) [reply keadm/cmd/keadm/app/cmd/helm/cloudcore.go]: Done, thanks

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go (hunk 2 above)
Comment (WillardHu) [reply keadm/cmd/keadm/app/cmd/helm/cloudcore.go]: Done, thanks

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go (hunk 3 above)
Comment (WillardHu) [reply keadm/cmd/keadm/app/cmd/helm/cloudcore.go]: Done, thanks

File: keadm/cmd/keadm/app/cmd/edge/upgrade.go (hunk 6 above)
Comment (WillardHu) [reply keadm/cmd/keadm/app/cmd/edge/upgrade.go]: Done, thanks

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go (hunk 7 above)
Comment (WillardHu) [reply keadm/cmd/keadm/app/cmd/helm/cloudcore.go]: Done, thanks

**Comment by WillardHu** [comment]
> I noticed that some of the code was not debugged; perhaps I should keep reviewing it once you have debugged it.

Ok, now this PR is for drawing out the design idea. There are still some issues to discuss at the next regular meeting. I will add unit tests later, and execute the cloud upgrade successfully.

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go (hunk 9 above)
Comment (WillardHu) [reply keadm/cmd/keadm/app/cmd/helm/cloudcore.go:73]: Ok, this function is used to abstract the helm related operations of the cloudcore, and the code of the cloud init has not been modified.

**Comment by WillardHu** [comment]
**TEST DATA**
```
[root@kubeedge-master ~]# helm list -n kubeedge
NAME     	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART          	APP VERSION
cloudcore	kubeedge 	1       	2023-12-14 16:07:26.729398813 +0800 CST	deployed	cloudcore-0.1.0	1.12.0

[root@kubeedge-master ~]# keadm upgrade cloud --help
Upgrade the cloud components to the desired version, it uses helm to upgrade the installed release of cloudcore.

Usage:
  keadm upgrade cloud [flags]

Flags:
      --advertise-address string    Please set the same value as when you installed it, this value is only used to generate the configuration and does not regenerate the certificate. eg: 10.10.102.78,10.10.102.79
  -d, --dry-run                     Print the generated k8s resources on the stdout, not actual execute. Always use in debug mode
      --external-helm-root string   Add external helm root path to keadm
      --force                       Forced upgrading the cloud components without waiting
  -h, --help                        help for cloud
      --kube-config string          Use this key to update kube-config path, eg: $HOME/.kube/config (default "/root/.kube/config")
      --kubeedge-version string     Use this key to set the upgrade image tag
      --print-final-values          Print the final values configuration for debuging
      --profile string              Sets profile on the command line. If '--values' is specified, this is ignored
      --recreate-pods               performs pods restart for the resource if applicable
      --reuse-values                reuse the last release's values and merge in any overrides from the command line via --set and -f.
      --set stringArray             Sets values on the command line (can specify multiple or separate values with commas: key1=val1,key2=val2)
      --values stringArray          specify values in a YAML file (can specify multiple)

[root@kubeedge-master ~]# keadm upgrade cloud --advertise-address=10.2.129.11 --kubeedge-version=v1.15.1 --recreate-pods --print-final-values
Kubernetes version verification passed, KubeEdge upgradation will start...
CLOUDCORE started
=========CHART DETAILS=======
Name: cloudcore
LAST DEPLOYED: Thu Dec 14 16:09:36 2023
NAMESPACE: kubeedge
STATUS: deployed
REVISION: 2
FINAL VALUES:
cloudCore:
  image:
    tag: v1.15.1
  modules:
    cloudHub:
      advertiseAddress:
      - 10.2.129.11
controllerManager:
  image:
    tag: v1.15.1
iptablesManager:
  image:
    tag: v1.15.1
This is cloudcore configuration of the previous version.
If you want to revert configuration items, please manually modify the configmap 'cloudcore'
and restart the cloudcore:
apiVersion: cloudcore.config.kubeedge.io/v1alpha2
commonConfig:
  monitorServer:
    bindAddress: 127.0.0.1:9091
  tunnelPort: 10351
kind: CloudCore
kubeAPIConfig:
  burst: 10000
  contentType: application/vnd.kubernetes.protobuf
  kubeConfig: ""
  master: ""
  qps: 5000
modules:
  cloudHub:
    advertiseAddress:
    - 10.2.129.11
    dnsNames:
    - ""
    edgeCertSigningDuration: 365
    enable: true
    https:
      address: 0.0.0.0
      enable: true
      port: 10002
    keepaliveInterval: 30
    nodeLimit: 1000
    quic:
      address: 0.0.0.0
      enable: false
      maxIncomingStreams: 10000
      port: 10001
    tlsCAFile: /etc/kubeedge/ca/rootCA.crt
    tlsCAKeyFile: /etc/kubeedge/ca/rootCA.key
    tlsCertFile: /etc/kubeedge/certs/edge.crt
    tlsPrivateKeyFile: /etc/kubeedge/certs/edge.key
    tokenRefreshDuration: 12
    unixsocket:
      address: unix:///var/lib/kubeedge/kubeedge.sock
      enable: true
    websocket:
      address: 0.0.0.0
      enable: true
      port: 10000
    writeTimeout: 30
  cloudStream:
    enable: true
    streamPort: 10003
    tlsStreamCAFile: /etc/kubeedge/ca/streamCA.crt
    tlsStreamCertFile: /etc/kubeedge/certs/stream.crt
    tlsStreamPrivateKeyFile: /etc/kubeedge/certs/stream.key
    tlsTunnelCAFile: /etc/kubeedge/ca/rootCA.crt
    tlsTunnelCertFile: /etc/kubeedge/certs/server.crt
    tlsTunnelPrivateKeyFile: /etc/kubeedge/certs/server.key
    tunnelPort: 10004
  deviceController:
    buffer:
      deviceEvent: 1
      deviceModelEvent: 1
      updateDeviceStatus: 1024
    enable: true
    load:
      updateDeviceStatusWorkers: 1
  dynamicController:
    enable: false
  edgeController:
    buffer:
      configMapEvent: 1
      createLease: 2024
      createNode: 1024
      deletePod: 1024
      patchNode: 1524
      patchPod: 1024
      podEvent: 1
      queryConfigMap: 1024
      queryLease: 1024
      queryNode: 2024
      queryPersistentVolume: 1024
      queryPersistentVolumeClaim: 1024
      querySecret: 1024
      queryVolumeAttachment: 1024
      ruleEndpointsEvent: 1
      rulesEvent: 1
      secretEvent: 1
      serviceAccountToken: 1024
      updateNode: 1024
      updateNodeStatus: 1024
      updatePodStatus: 1024
    enable: true
    load:
      ServiceAccountTokenWorkers: 100
      UpdateRuleStatusWorkers: 4
      createLeaseWorkers: 1000
      createNodeWorkers: 100
      deletePodWorkers: 100
      patchNodeWorkers: 120
      patchPodWorkers: 100
      queryConfigMapWorkers: 100
      queryLeaseWorkers: 100
      queryNodeWorkers: 1000
      queryPersistentVolumeClaimWorkers: 4
      queryPersistentVolumeWorkers: 4
      querySecretWorkers: 100
      queryVolumeAttachmentWorkers: 4
      updateNodeStatusWorkers: 1
      updateNodeWorkers: 4
      updatePodStatusWorkers: 1
    nodeUpdateFrequency: 10
  iptablesManager:
    enable: true
    mode: internal
  nodeUpgradeJobController:
    buffer:
      nodeUpgradeJobEvent: 1
      updateNodeUpgradeJobStatus: 1024
    enable: false
    load:
      nodeUpgradeJobWorkers: 1
  router:
    address: 0.0.0.0
    enable: false
    port: 9443
    restTimeout: 60
  syncController:
    enable: true

[root@kubeedge-master ~]# helm list -n kubeedge
NAME     	NAMESPACE	REVISION	UPDATED                               	STATUS  	CHART           	APP VERSION
cloudcore	kubeedge 	2       	2023-12-14 16:09:36.67522936 +0800 CST	deployed	cloudcore-1.15.1	1.15.1

[root@kubeedge-master ~]# kubectl get po -n kubeedge
NAME                         READY   STATUS    RESTARTS   AGE
cloudcore-5ff67dd45f-sztn2   1/1     Running   0          2m15s
```

**Comment by WillardHu** [comment]
PTAL; @Shelley-BaoYue @fisherxu @ZhengXinwei-F

**Comment by WillardHu** [comment]
> Could you modify this PR based my PR? #5330

Okay, I'll wait for your PR to merge first.

**Comment by WillardHu** [comment]
Rebase the PR https://github.com/kubeedge/kubeedge/pull/5330, and submit a new commit to fix the edge upgrade command.

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go [hunk 10]
```diff
@@ -0,0 +1,325 @@
+/*
+Copyright 2019 The KubeEdge Authors.
```
Comment (fisherxu) [inline keadm/cmd/keadm/app/cmd/helm/cloudcore.go:2]: 2023 or 24?

File: keadm/cmd/keadm/app/cmd/cloud/upgrade.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/helm"
+)
+
+func NewCloudUpgrade() *cobra.Command {
+	opts := newCloudUpgradeOptions()
+
+	cmd := &cobra.Command{
+		Use:   "cloud",
+		Short: "Upgrade the cloud components",
+		Long:  "Upgrade the cloud components to the desired version, it uses helm to upgrade the installed release of cloudcore.",
```
Comment (fisherxu) [inline keadm/cmd/keadm/app/cmd/cloud/upgrade.go:16]: > it uses helm to upgrade the installed release of cloudcore.

Other components should also use helm to upgrade?

File: keadm/cmd/keadm/app/cmd/cloud/upgrade.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
+package cloud
```
[the remark this answers is not in the document]
Comment (WillardHu) [reply keadm/cmd/keadm/app/cmd/cloud/upgrade.go:16]: Done, thanks.

File: keadm/cmd/keadm/app/cmd/helm/cloudcore_test.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
+package helm
```
[the remark this answers is not in the document]
Comment (WillardHu) [reply keadm/cmd/keadm/app/cmd/helm/cloudcore_test.go:16]: Done, thanks.

File: keadm/cmd/keadm/app/cmd/helm/cloudcore.go (hunk 10 above)
Comment (WillardHu) [reply keadm/cmd/keadm/app/cmd/helm/cloudcore.go]: I added the license header to all new files and changed the date to 2024.

File: keadm/cmd/keadm/app/cmd/cloud/upgrade.go (hunk 11 above)
Comment (WillardHu) [reply keadm/cmd/keadm/app/cmd/cloud/upgrade.go:31]: Yes, cloud components cloudcore, controllermanager, admission and iptablemanager are all in the cloudcore chart.

Comment (fisherxu) [reply keadm/cmd/keadm/app/cmd/cloud/upgrade.go:31]:
```suggestion
		Long:  "Upgrade the cloud components to the desired version, it uses helm to upgrade the installed release of cloudcore chart, which includes all the cloud components.",
```
Okey, I think we can add some description for the cloudcore chart, or user will be confused, think we only upgrade the cloudcore.

Comment (WillardHu) [reply keadm/cmd/keadm/app/cmd/cloud/upgrade.go]: Done, thanks.

**Review by fisherxu — COMMENTED** [review]
Now let's get the first version of cloudcore upgrade merged, and fix any bugs in follow-up PRs.
/lgtm
/approve

**Pull request merged**